library/Test/Hspec/Expectations/Json/Lifted.hs

```diff
@@ -2,6 +2,7 @@ module Test.Hspec.Expectations.Json.Lifted
   ( shouldBeJson
   , shouldBeUnorderedJson
   , shouldMatchJson
+  , shouldMatchJsonWithOmittedNullFields
   , shouldMatchOrderedJson
```

Comment:
What do you think _not_ making more distinct functions? This name is quite out of hand, and we seem to be wanting to fine-tune this behavior more than once so I'd rather do something more scalable.

My suggestion would be to introduce an options type, implement the 4 functions we have in terms of it, then provide a `With` function that can be used for any other permutations.

```hs
data MatchOptions = MatchOptions
  { tolerateMissingFields :: Bool
  , enforceArrayOrdering :: Bool
  }

defaultMatchOptions = MatchOptions
  { tolerateMissingFields = True
  , enforceArrayOrdering = False
  }

shouldMatchJson = shouldMatchJsonWith defaultMatchOptions

shouldMatchOrderedJson = shouldMatchJSONWith $ defaultMatchOptions
  { enforceArrayOrder = True
  }

shouldBeUnorderedJson = shouldMatchJsonWith $ defaultMatchOptions
  { tolerateMissingFields = False
  }

shouldBeJson = shouldMatchJsonWith $ defaultMatchOptions
  { tolerateMissingFields = False
  , enforceArrayOrder = True
  }
```

Then you can introduce `MatchOptions{treatNullAsMissing}` or similar. I would let the default of `shouldMatch*` be `True`, I don't see much of a downside, then let folks drop down to `shouldMatchJsonWith` to be stricter.

Reply:
I like the idea of making the surface area more tenable with `MatchOptions`, but that would also require supporting every 2^3 combination of option which I'd like to avoid. 

This new function really builds on `shouldMatchJson` and would require different strategies to implement the other combinations.

Reply:
> but that would also require supporting every 2^3 combination of option which I'd like to avoid

Hmm, I don't follow.  I was thinking a single `shouldMatchJsonWith {options}`, where `options` has `n` fields, would be exactly how to _avoid_ making 2^n new `should{Match|Be}(Unordered)(WithThisOrThat)` functions. But you make it sound like it introduces the problem instead of fixing it.

Expanding from this PR as-is to support all variations would have to add:

```hs
shouldMatchOrderedJsonWithOmittedNullFields
shouldBeUnorderedJsonWithOmittedNullFields
shouldBeJsonWithOmittedNullFields
```

Where-as my suggestion would introduce no functions at all.

Reply:
I think I am misunderstanding your suggestion then, is `MatchOptions`  and `shouldMatchJsonWith :: MatchOptions -> ...` not something exposed to the client?

Reply:
It is, and is 1 function that prevents having to implement 2^n functions.

Reply:
We would still need to handle 2^n permutations is what I was getting at- right now we only have an implementation that works for the 5 functions that this module exports. 

My point was that adding the other 3 is not straightforward as it will require different techniques.

Reply:
I gave `MatchOptions` a go and think it actually looks pretty nice. Turns out thinking about this as applying different filtering passes did make this a lot simpler.

I still did have to do the nasty `normalizeArrays`, but I think we have to do something when array elements are not the same shape and this seems to be as valid a thing to do as the other algorithm.